### PR TITLE
#9469 Refactor Plugin Interface

### PR DESCRIPTION
**Pull request by BrianLondon**
Fixes [Hide implementation details from nu-plugin public interface](https://github.com/nushell/nushell/issues/9469)

# Description
This change makes some changes to the plugin API with the intent of better encapsulating implementation details as well as better partitioning those parts of the `nu-plugin` crate to separate public items used by plugin authors from public items used by other Nushell crates.

# User-Facing Changes

This is a breaking change for existing Plugins. All plugin samples in this repo have been updated.

Instead of serving a plugin like this:
```rust
fn main() {
   serve_plugin(&mut MyPlugin{}, MsgPackSerializer)
}
```

a plugin author must now:
```rust
fn main() {
   serve_plugin(&mut MyPlugin{}, EncodingType::MsgPack)
}
```

Some items that were never meant for plugin authors are now behind a crate feature in `nu-plugin`. This feature is activated in `crates/nu-parser/Cargo.toml`. They have also been moved to a sub-module which allows us to add documentation while avoiding confusion over whether certain items should be used outside of Nushell internals.

# Tests + Formatting
- :green_circle: `toolkit fmt`
- :green_circle: `toolkit clippy`
- :green_circle: `toolkit test`
- :green_circle: `toolkit test stdlib`

# After Submitting

This change will require an update to the hidden plugin tutorial.

### CODE REVIEW

**Comment by BrianLondon** [comment]
I realize I could have been more clear above. When I said:

> This is a breaking change for existing Plugins

I mean that existing plugins will require a code change if they upgrade to the current version of `nu-plugin`. This should not break existing plugins ability to interface the current version of `nu`.

File: crates/nu-plugin/src/plugin/mod.rs [hunk 1]
```diff
@@ -265,16 +269,23 @@ pub trait Plugin {
...
 /// The object that is expected to be received by nushell is the `PluginResponse` struct.
 /// The `serve_plugin` function should ensure that it is encoded correctly and sent
 /// to StdOut for nushell to decode and and present its result.
-pub fn serve_plugin(plugin: &mut impl Plugin, encoder: impl PluginEncoder) {
+pub fn serve_plugin(plugin: &mut impl Plugin, encoder_format: EncodingType) {
+    match encoder_format {
+        EncodingType::Json => serve_plugin_with_format::<JsonSerializer>(plugin),
+        EncodingType::MsgPack => serve_plugin_with_format::<MsgPackSerializer>(plugin),
+    }
+}
```
Comment (sholderbach) [inline crates/nu-plugin/src/plugin/mod.rs:284]: Mhh, I think there is a clear potential downside to the approach of using an enum here.

On the crate boundary this function will be compiled momorphizing `serve_plugin_with_format::<T: PluginEncoder>` for all options and using all variants of `serve_plugin_with_format` in the match, even though a plugin implementation using `nu-plugin` will typically only use one `PluginEncoder`/`EncodingType` variant. To get rid of the result of all this codegen (the serdes are big!) the dead-code elimination needs to constant fold across the crate boundary, which requires goodwill by LLVM or succesfull LTO. 

I would like to see the before and after comparison with `cargo bloat`/`cargo build --timings` for a plugin.
On the nushell side itself we won't be able to get rid of having the serde code for all formats we support, but we should be nice to the plugin authors in that regard.

Comment (BrianLondon) [reply crates/nu-plugin/src/plugin/mod.rs:284]: Thank you for the comment.

Just to make sure I understand your concern: is your objection that by referencing both serializers, plugins will bundle both serialization libraries thereby increasing the compiled binary size of the plugin? Or are there other concerns with monomophizing `serve_plugin_with_format for` both encoders?

Comment (sholderbach) [reply crates/nu-plugin/src/plugin/mod.rs:284]: Yes the direct concern is just code/binary size and any downstream effects from that.

This adds close to 0.8 MB or 15% to `nu_plugin_formats` 

## Output from `cargo bloat -p nu_plugin_formats --release`

### This branch
Total: 6.1 MiB
`.text`: 2.7 MiB
<details>
   <summary>full output (function level breakdown)</summary>

```
    Analyzing target/release/nu_plugin_formats

 File  .text    Size        Crate Name
 0.6%   1.3% 35.9KiB        regex <regex::exec::ExecNoSync as regex::re_trait::RegularExpression>::captures_read_at
 0.3%   0.8% 21.7KiB          std addr2line::ResDwarf<R>::parse
 0.3%   0.7% 18.7KiB          std std::backtrace_rs::symbolize::gimli::resolve::{{closure}}
 0.3%   0.6% 16.1KiB    nu_plugin <nu_protocol::shell_error::_::<impl serde::de::Deserialize for nu_protocol::shell_error::ShellError>::deserialize::__FieldVisitor as serde::de::Visitor>::visit_bytes
 0.2%   0.5% 15.0KiB   rmp_serde? <&mut rmp_serde::decode::Deserializer<R,C> as serde::de::Deserializer>::deserialize_any
 0.2%   0.5% 12.9KiB        regex regex::exec::ExecBuilder::build
 0.2%   0.4% 10.6KiB          std addr2line::ResUnit<R>::parse_lines
 0.2%   0.4%  9.9KiB  serde_json? <&mut serde_json::de::Deserializer<R> as serde::de::Deserializer>::deserialize_struct
 0.1%   0.3%  9.2KiB    nu_plugin nu_plugin::plugin::serve_plugin
 0.1%   0.3%  9.0KiB   eml_parser eml_parser::parser::EmlParser::parse
 0.1%   0.3%  8.2KiB  nu_protocol <alloc::vec::Vec<T,A> as core::clone::Clone>::clone
 0.1%   0.3%  8.1KiB          std miniz_oxide::inflate::core::decompress
 0.1%   0.3%  7.6KiB regex_syntax <regex_syntax::hir::translate::TranslatorI as regex_syntax::ast::visitor::Visitor>::visit_post
 0.1%   0.2%  6.8KiB          std <T as alloc::alloc::WriteCloneIntoRaw>::write_clone_into_raw
 0.1%   0.2%  6.6KiB aho_corasick aho_corasick::automaton::try_find_fwd
 0.1%   0.2%  6.6KiB   rmp_serde? <&mut rmp_serde::decode::Deserializer<R,C> as serde::de::Deserializer>::deserialize_any
 0.1%   0.2%  6.3KiB   rmp_serde? <&mut rmp_serde::decode::Deserializer<R,C> as serde::de::Deserializer>::deserialize_any
 0.1%   0.2%  6.2KiB   rmp_serde? <&mut rmp_serde::decode::Deserializer<R,C> as serde::de::Deserializer>::deserialize_any
 0.1%   0.2%  5.9KiB    nu_plugin nu_protocol::shell_error::_::<impl serde::ser::Serialize for nu_protocol::shell_error::ShellError>::serialize
 0.1%   0.2%  5.9KiB   rmp_serde? <&mut rmp_serde::decode::Deserializer<R,C> as serde::de::Deserializer>::deserialize_any
40.5%  91.7%  2.5MiB              And 8252 smaller methods. Use -n N to show more.
44.2% 100.0%  2.7MiB              .text section size, the file size is 6.1MiB
```
</details>

### `main` @ `a0dafcb0f` (effectively the base commit, deviation by one commit in a separate crate)

Total: 5.3 MiB
`.text`: 2.3 MiB
<details>
   <summary>full output (function level breakdown)</summary>
```
    Analyzing target/release/nu_plugin_formats

 File  .text    Size        Crate Name
 0.7%   1.5% 35.9KiB        regex <regex::exec::ExecNoSync as regex::re_trait::RegularExpression>::captures_read_at
 0.4%   0.9% 21.7KiB          std addr2line::ResDwarf<R>::parse
 0.3%   0.8% 18.7KiB          std std::backtrace_rs::symbolize::gimli::resolve::{{closure}}
 0.3%   0.7% 16.1KiB    nu_plugin <nu_protocol::shell_error::_::<impl serde::de::Deserialize for nu_protocol::shell_error::ShellError>::deserialize::__FieldVisitor as serde::de::Visitor>::visit_bytes
 0.3%   0.6% 15.0KiB   rmp_serde? <&mut rmp_serde::decode::Deserializer<R,C> as serde::de::Deserializer>::deserialize_any
 0.2%   0.6% 12.9KiB        regex regex::exec::ExecBuilder::build
 0.2%   0.5% 10.6KiB          std addr2line::ResUnit<R>::parse_lines
 0.2%   0.4%  9.0KiB   eml_parser eml_parser::parser::EmlParser::parse
 0.2%   0.4%  8.2KiB  nu_protocol <alloc::vec::Vec<T,A> as core::clone::Clone>::clone
 0.1%   0.3%  8.1KiB          std miniz_oxide::inflate::core::decompress
 0.1%   0.3%  7.6KiB regex_syntax <regex_syntax::hir::translate::TranslatorI as regex_syntax::ast::visitor::Visitor>::visit_post
 0.1%   0.3%  6.8KiB   rmp_serde? <&mut rmp_serde::decode::Deserializer<R,C> as serde::de::Deserializer>::deserialize_any
 0.1%   0.3%  6.8KiB          std <T as alloc::alloc::WriteCloneIntoRaw>::write_clone_into_raw
 0.1%   0.3%  6.6KiB aho_corasick aho_corasick::automaton::try_find_fwd
 0.1%   0.3%  6.3KiB   rmp_serde? <&mut rmp_serde::decode::Deserializer<R,C> as serde::de::Deserializer>::deserialize_any
 0.1%   0.3%  6.2KiB   rmp_serde? <&mut rmp_serde::decode::Deserializer<R,C> as serde::de::Deserializer>::deserialize_any
 0.1%   0.3%  5.9KiB    nu_plugin nu_protocol::shell_error::_::<impl serde::ser::Serialize for nu_protocol::shell_error::ShellError>::serialize
 0.1%   0.3%  5.9KiB   rmp_serde? <&mut rmp_serde::decode::Deserializer<R,C> as serde::de::Deserializer>::deserialize_any
 0.1%   0.3%  5.9KiB  nu_protocol nu_protocol::shell_error::_::<impl serde::ser::Serialize for nu_protocol::shell_error::ShellError>::serialize
 0.1%   0.3%  5.9KiB  nu_protocol <nu_protocol::shell_error::ShellError as core::fmt::Debug>::fmt
38.6%  90.3%  2.0MiB              And 7000 smaller methods. Use -n N to show more.
42.8% 100.0%  2.3MiB              .text section size, the file size is 5.3MiB
```
</details>

**Review by sholderbach — CHANGES_REQUESTED** [review]
As it significantly worsens the binary size of plugins compiled using `nu-plugin` we should probably revert from the `EncodingType` enum to the generic `serve_plugin()` implementation.

As part of your changes you converted the [object safe](https://doc.rust-lang.org/reference/items/traits.html#object-safety) `trait PluginEncoder` from which `dyn` trait objects could be constructed to a basic trait. I don't think that the ability to hot-swap the `PluginEncoder` through trait objects is really needed in practice. (The only theoretical usecase I could imagine that is not exercised in practice would be easy mocking in some tests.) So I am neutral to that part of the changes.

But the change to split the components that are plugin relevant from the other components that are only needed on the `nu` side has a measurable benefit on plugin compile times. So I think we should take this idea.
One possible idea to take this even further would be to make the two encodings also features as a plugin will only need to be built with one. Then we would not only save a bit of compile time but also possilby a few dependencies in the plugin compile tree.

## Binary size: `cargo bloat -p nu_plugin_formats --release`

https://github.com/RazrFalcon/cargo-bloat

### This branch
Total: 6.1 MiB
`.text`: 2.7 MiB
<details>
   <summary>full output (function level breakdown)</summary>

```
    Analyzing target/release/nu_plugin_formats

 File  .text    Size        Crate Name
 0.6%   1.3% 35.9KiB        regex <regex::exec::ExecNoSync as regex::re_trait::RegularExpression>::captures_read_at
 0.3%   0.8% 21.7KiB          std addr2line::ResDwarf<R>::parse
 0.3%   0.7% 18.7KiB          std std::backtrace_rs::symbolize::gimli::resolve::{{closure}}
 0.3%   0.6% 16.1KiB    nu_plugin <nu_protocol::shell_error::_::<impl serde::de::Deserialize for nu_protocol::shell_error::ShellError>::deserialize::__FieldVisitor as serde::de::Visitor>::visit_bytes
 0.2%   0.5% 15.0KiB   rmp_serde? <&mut rmp_serde::decode::Deserializer<R,C> as serde::de::Deserializer>::deserialize_any
 0.2%   0.5% 12.9KiB        regex regex::exec::ExecBuilder::build
 0.2%   0.4% 10.6KiB          std addr2line::ResUnit<R>::parse_lines
 0.2%   0.4%  9.9KiB  serde_json? <&mut serde_json::de::Deserializer<R> as serde::de::Deserializer>::deserialize_struct
 0.1%   0.3%  9.2KiB    nu_plugin nu_plugin::plugin::serve_plugin
 0.1%   0.3%  9.0KiB   eml_parser eml_parser::parser::EmlParser::parse
 0.1%   0.3%  8.2KiB  nu_protocol <alloc::vec::Vec<T,A> as core::clone::Clone>::clone
 0.1%   0.3%  8.1KiB          std miniz_oxide::inflate::core::decompress
 0.1%   0.3%  7.6KiB regex_syntax <regex_syntax::hir::translate::TranslatorI as regex_syntax::ast::visitor::Visitor>::visit_post
 0.1%   0.2%  6.8KiB          std <T as alloc::alloc::WriteCloneIntoRaw>::write_clone_into_raw
 0.1%   0.2%  6.6KiB aho_corasick aho_corasick::automaton::try_find_fwd
 0.1%   0.2%  6.6KiB   rmp_serde? <&mut rmp_serde::decode::Deserializer<R,C> as serde::de::Deserializer>::deserialize_any
 0.1%   0.2%  6.3KiB   rmp_serde? <&mut rmp_serde::decode::Deserializer<R,C> as serde::de::Deserializer>::deserialize_any
 0.1%   0.2%  6.2KiB   rmp_serde? <&mut rmp_serde::decode::Deserializer<R,C> as serde::de::Deserializer>::deserialize_any
 0.1%   0.2%  5.9KiB    nu_plugin nu_protocol::shell_error::_::<impl serde::ser::Serialize for nu_protocol::shell_error::ShellError>::serialize
 0.1%   0.2%  5.9KiB   rmp_serde? <&mut rmp_serde::decode::Deserializer<R,C> as serde::de::Deserializer>::deserialize_any
40.5%  91.7%  2.5MiB              And 8252 smaller methods. Use -n N to show more.
44.2% 100.0%  2.7MiB              .text section size, the file size is 6.1MiB
```
</details>

### `main` @ `a0dafcb0f` (effectively the base commit, deviation by one commit in a separate crate)

Total: 5.3 MiB
`.text`: 2.3 MiB
<details>
   <summary>full output (function level breakdown)</summary>

```
    Analyzing target/release/nu_plugin_formats

 File  .text    Size        Crate Name
 0.7%   1.5% 35.9KiB        regex <regex::exec::ExecNoSync as regex::re_trait::RegularExpression>::captures_read_at
 0.4%   0.9% 21.7KiB          std addr2line::ResDwarf<R>::parse
 0.3%   0.8% 18.7KiB          std std::backtrace_rs::symbolize::gimli::resolve::{{closure}}
 0.3%   0.7% 16.1KiB    nu_plugin <nu_protocol::shell_error::_::<impl serde::de::Deserialize for nu_protocol::shell_error::ShellError>::deserialize::__FieldVisitor as serde::de::Visitor>::visit_bytes
 0.3%   0.6% 15.0KiB   rmp_serde? <&mut rmp_serde::decode::Deserializer<R,C> as serde::de::Deserializer>::deserialize_any
 0.2%   0.6% 12.9KiB        regex regex::exec::ExecBuilder::build
 0.2%   0.5% 10.6KiB          std addr2line::ResUnit<R>::parse_lines
 0.2%   0.4%  9.0KiB   eml_parser eml_parser::parser::EmlParser::parse
 0.2%   0.4%  8.2KiB  nu_protocol <alloc::vec::Vec<T,A> as core::clone::Clone>::clone
 0.1%   0.3%  8.1KiB          std miniz_oxide::inflate::core::decompress
 0.1%   0.3%  7.6KiB regex_syntax <regex_syntax::hir::translate::TranslatorI as regex_syntax::ast::visitor::Visitor>::visit_post
 0.1%   0.3%  6.8KiB   rmp_serde? <&mut rmp_serde::decode::Deserializer<R,C> as serde::de::Deserializer>::deserialize_any
 0.1%   0.3%  6.8KiB          std <T as alloc::alloc::WriteCloneIntoRaw>::write_clone_into_raw
 0.1%   0.3%  6.6KiB aho_corasick aho_corasick::automaton::try_find_fwd
 0.1%   0.3%  6.3KiB   rmp_serde? <&mut rmp_serde::decode::Deserializer<R,C> as serde::de::Deserializer>::deserialize_any
 0.1%   0.3%  6.2KiB   rmp_serde? <&mut rmp_serde::decode::Deserializer<R,C> as serde::de::Deserializer>::deserialize_any
 0.1%   0.3%  5.9KiB    nu_plugin nu_protocol::shell_error::_::<impl serde::ser::Serialize for nu_protocol::shell_error::ShellError>::serialize
 0.1%   0.3%  5.9KiB   rmp_serde? <&mut rmp_serde::decode::Deserializer<R,C> as serde::de::Deserializer>::deserialize_any
 0.1%   0.3%  5.9KiB  nu_protocol nu_protocol::shell_error::_::<impl serde::ser::Serialize for nu_protocol::shell_error::ShellError>::serialize
 0.1%   0.3%  5.9KiB  nu_protocol <nu_protocol::shell_error::ShellError as core::fmt::Debug>::fmt
38.6%  90.3%  2.0MiB              And 7000 smaller methods. Use -n N to show more.
42.8% 100.0%  2.3MiB              .text section size, the file size is 5.3MiB
```
</details>

## Compile times: `cargo build -p nu_plugin_formats --timings`
To compile from scratch run `cargo clean --profile dev` before

### This branch
grafik


### `main`
grafik

**Comment by BrianLondon** [comment]
Alright, my goal is still to minimize the leakiness of the internal serialization details to the plugin authors. Rather than just rolling back that half I'm going to work a bit to accomplish both that and the JSON library exclusion for binary size considerations. I'm going to play with it a bit to see if I can get both without adding any more feature flags.

**Comment by sholderbach** [comment]
> Even with the trait approach I think we can do it without dynamic dispatch or instantiating a marker object.

Agree we don't need dynamic dispatch here. So I am fine with you removing the `fn ...(&self,` bits from the trait.


> Alright, my goal is still to minimize the leakiness of the internal serialization details to the plugin authors.

That is certainly a noble goal. Limiting us so we don't make semver incompatible changes on the semi internal level would be great. But we don't follow full compatible semver versioning on the individual crate level yet. We update the whole crate workspace even though subcrates may remain compatible (causing churn in version resolution)

The priority for us should be that we avoid causing problems in the happy path so that a change doesn't break stuff without clear resolution. Blocking all possible intentional abuse of our API will have to remain second priority.

> Rather than just rolling back that half I'm going to work a bit to accomplish both that and the JSON library exclusion for binary size considerations. I'm going to play with it a bit to see if I can get both without adding any more feature flags.

Cool! Thanks. Yeah feature flags get iffy quickly especially around testing.

**Comment by sholderbach** [comment]
Thanks for your efforts and investigation here.

The plugin interface has been superseded by @devyn 's work to make the plugin protocol both more flexible and robust.

The general idea of #9469 is still valid that user-facing `nu-plugin` and engine facing logic should be separated if possible.

**Pull request closed without merging**